conflation/map_matching.py

```diff
@@ -0,0 +1,292 @@
+        )
+
+    finished_bbox_sections = multiprocessing.Value("i", 0)
+    with multiprocessing.Pool(
+        initializer=initialize_multiprocess,
+        initargs=(map_matches_dir, valhalla_url, valhalla_headers, finished_bbox_sections),
+        processes=processes,
+    ) as pool:
+        result = pool.map_async(map_match_for_bbox, bbox_sections)
+
+        progress = 0
+        increment = 5
+        while not result.ready():
+            result.wait(timeout=5)
+            next_progress = int(finished_bbox_sections.value / len(bbox_sections) * 100)
+            if int(next_progress / increment) > progress:
+                print("Current progress: {}%".format(next_progress))
+                progress = int(next_progress / increment)
+        if progress != 100 / increment:
+            print("Current progress: 100%")
+
+
+def initialize_multiprocess(
+    global_map_matches_dir_: str,
+    global_valhalla_url_: str,
+    global_valhalla_headers_: dict,
+    finished_bbox_sections_: multiprocessing.Value,
+) -> None:
+    """
+    Initializes global variables referenced / updated by all threads of the multiprocess map matching requests.
+    """
+
+    global global_map_matches_dir
+    global_map_matches_dir = global_map_matches_dir_
+
+    # Integer counter of num of finished bbox_sections
+    global finished_bbox_sections
+    finished_bbox_sections = finished_bbox_sections_
+
+    # Valhalla base url and headers
+    global global_valhalla_url
+    global_valhalla_url = global_valhalla_url_
+    global global_valhalla_headers
+    global_valhalla_headers = global_valhalla_headers_
+
+
+def map_match_for_bbox(bbox_section: tuple[str, str]) -> None:
+    """
+    Performs map matching for a given bbox section. It pulls the bbox section's traces from the previous steps by
+    finding it on the filesystem. For each sequence in the traces, it passes it to add_map_matches_for_shape to make the
+    actual Valhalla call. Map matching results are propagated to disk with write_map_matches and the original traces
+    pickle is renamed as a checkpoint.
+
+    :param bbox_section: From util.split_bbox, used in previous step
+    """
+    try:
+        bbox, trace_filename = bbox_section
+        processed_trace_filename = util.get_processed_trace_filename(trace_filename)
+
+        # Check to see if the trace has already been processed by map_matching
+        if os.path.exists(processed_trace_filename):
+            print("Map matching already complete for bbox={}. Skipping...".format(bbox))
+            with finished_bbox_sections.get_lock():
+                finished_bbox_sections.value += 1
+            return
+
+        trace_data: list[list[dict]] = pickle.load(open(trace_filename, "rb"))
+        map_matches = {}
+        [add_map_matches_for_shape(map_matches, shape) for shape in trace_data]
+        if len(map_matches):
+            write_map_matches(global_map_matches_dir, map_matches)
```

Comment:
Creates an empty `map_matches` dict and `add_map_matches_for_shape` adds to it in place. Then `write_map_matches` writes the dict to disk.